File: filesystem/src/test/java/com/nytimes/android/external/fs3/StoreNetworkBeforeStaleFailTest.java
package com.nytimes.android.external.fs3;

import com.nytimes.android.external.store3.base.Fetcher;
import com.nytimes.android.external.store3.base.Persister;
import com.nytimes.android.external.store3.base.RecordProvider;
import com.nytimes.android.external.store3.base.RecordState;
import com.nytimes.android.external.store3.base.impl.BarCode;
import com.nytimes.android.external.store3.base.impl.Store;
import com.nytimes.android.external.store3.base.impl.StoreBuilder;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.runners.MockitoJUnitRunner;

import javax.annotation.Nonnull;

import io.reactivex.Maybe;
import io.reactivex.Single;
import okio.BufferedSource;

import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class StoreNetworkBeforeStaleFailTest {
    static final Exception sorry = new Exception("sorry");
    private static final BarCode barCode = new BarCode("key", "value");
    @Mock
    Fetcher<BufferedSource, BarCode> fetcher;
    Store<BufferedSource, BarCode> store;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        store = StoreBuilder.<BufferedSource>barcode()
                .fetcher(fetcher)
                .persister(new TestPersister())
                .networkBeforeStale()
                .open();

    }

    @Test
    public void networkBeforeStaleNoNetworkResponse() {
        Single<BufferedSource> exception = Single.error(sorry);
        when(fetcher.fetch(barCode))
                .thenReturn(exception);
        store.get(barCode).test().assertError(sorry);
        verify(fetcher, times(1)).fetch(barCode);
    }

    private static final class TestPersister implements Persister<BufferedSource, BarCode>, RecordProvider<BarCode> {
        @Nonnull
        @Override
        public RecordState getRecordState(@Nonnull BarCode barCode) {
            return RecordState.MISSING;
        }

        @Nonnull
        @Override
        public Maybe<BufferedSource> read(@Nonnull BarCode barCode) {
            return Maybe.error(sorry);
        }

        @Nonnull
        @Override
        public Single<Boolean> write(@Nonnull BarCode barCode,
                                         @Nonnull BufferedSource bufferedSource) {
            return Single.just(true);
        }
    }
}

File: filesystem/src/test/java/com/nytimes/android/external/fs3/StoreNetworkBeforeStaleFailTest.kt
package com.nytimes.android.external.fs3

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import com.nytimes.android.external.store3.base.Fetcher
import com.nytimes.android.external.store3.base.Persister
import com.nytimes.android.external.store3.base.RecordProvider
import com.nytimes.android.external.store3.base.RecordState
import com.nytimes.android.external.store3.base.impl.BarCode
import com.nytimes.android.external.store3.base.impl.StoreBuilder
import kotlinx.coroutines.runBlocking
import okio.BufferedSource
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.junit.Test
import org.mockito.Mockito.times
import org.mockito.Mockito.verify

class StoreNetworkBeforeStaleFailTest {
    private val fetcher: Fetcher<BufferedSource, BarCode> = mock()
    private val store = StoreBuilder.barcode<BufferedSource>()
            .fetcher(fetcher)
            .persister(TestPersister())
            .networkBeforeStale()
            .open()

    @Test
    fun networkBeforeStaleNoNetworkResponse() = runBlocking<Unit> {
        whenever(fetcher.fetch(barCode)).thenThrow(SORRY)
        try {
            store.get(barCode)
            fail("Exception not thrown")
        } catch (e: Exception) {
            assertThat(e.localizedMessage).isEqualTo(SORRY.localizedMessage)
        }
        verify(fetcher, times(1)).fetch(barCode)
    }

    private class TestPersister : Persister<BufferedSource, BarCode>, RecordProvider<BarCode> {
        override fun getRecordState(barCode: BarCode): RecordState {
            return RecordState.MISSING
        }

        override suspend fun read(key: BarCode): BufferedSource? {
            throw SORRY
        }

        override suspend fun write(key: BarCode, raw: BufferedSource) = true
    }

    companion object {
        private val SORRY = RuntimeException("sorry")
        private val barCode = BarCode("key", "value")
    }
}
